File: aws/CloudFormation/CognitoUserPoolDomain/cognitoUserPoolDomain.go
package main

import (
	"context"
	"github.com/aws/aws-lambda-go/cfn"
	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/aws/awserr"
	"github.com/aws/aws-sdk-go-v2/aws/external"
	"github.com/aws/aws-sdk-go-v2/service/cognitoidentityprovider"
	"github.com/codesmith-gmbh/cgc/cgccf"
	"github.com/codesmith-gmbh/forge/aws/common"
	"github.com/mitchellh/mapstructure"
	"github.com/pkg/errors"
	"strings"
)

func main() {
	p := newProc()
	cgccf.StartEventProcessor(p)
}

type proc struct {
	idp *cognitoidentityprovider.Client
}

func newProc() cgccf.EventProcessor {
	cfg, err := external.LoadDefaultAWSConfig()
	if err != nil {
		return &cgccf.ConstantErrorEventProcessor{Error: err}
	}
	return newProcFromConfig(cfg)
}

func newProcFromConfig(cfg aws.Config) *proc {
	return &proc{idp: cognitoidentityprovider.New(cfg)}
}

type Properties struct {
	Domain             string
	UserPoolId         string
	CustomDomainConfig cognitoidentityprovider.CustomDomainConfigType
}

func validateProperties(input map[string]interface{}) (Properties, error) {
	var properties Properties
	if err := mapstructure.Decode(input, &properties); err != nil {
		return properties, err
	}
	return properties, nil
}

func (p *proc) ProcessEvent(ctx context.Context, event cfn.Event) (string, map[string]interface{}, error) {
	properties, err := validateProperties(event.ResourceProperties)
	if err != nil {
		return "", nil, err
	}
	switch event.RequestType {
	case cfn.RequestDelete:
		return p.deleteDomain(ctx, event, properties)
	case cfn.RequestCreate, cfn.RequestUpdate:
		return p.createDomain(ctx, event, properties)
	default:
		return common.UnknownRequestType(event)
	}
}

func (p *proc) createDomain(ctx context.Context, event cfn.Event, properties Properties) (string, map[string]interface{}, error) {
	var out *cognitoidentityprovider.CreateUserPoolDomainResponse
	var err error
	if properties.CustomDomainConfig.CertificateArn == nil {
		out, err = p.idp.CreateUserPoolDomainRequest(&cognitoidentityprovider.CreateUserPoolDomainInput{
			Domain:     &properties.Domain,
			UserPoolId: &properties.UserPoolId,
		}).Send(ctx)
	} else {
		out, err = p.idp.CreateUserPoolDomainRequest(&cognitoidentityprovider.CreateUserPoolDomainInput{
			Domain:             &properties.Domain,
			UserPoolId:         &properties.UserPoolId,
			CustomDomainConfig: &properties.CustomDomainConfig,
		}).Send(ctx)
	}
	if err != nil {
		return "", nil, errors.Wrap(err, "Could not create the CognitoUserPoolDomain")
	}
	var cloudFrontDomain string
	var domain string
	if out.CloudFrontDomain == nil {
		cloudFrontDomain = ""
		domain = properties.Domain + ".auth." + p.idp.Region + ".amazoncognito.com"
	} else {
		cloudFrontDomain = *out.CloudFrontDomain
		domain = properties.Domain
	}
	return properties.Domain,
		map[string]interface{}{
			"UserPoolId":       properties.UserPoolId,
			"CloudFrontDomain": cloudFrontDomain,
			"Domain":           domain,
		},
		nil
}

func (p *proc) deleteDomain(ctx context.Context, event cfn.Event, properties Properties) (string, map[string]interface{}, error) {
	_, err := p.idp.DeleteUserPoolDomainRequest(&cognitoidentityprovider.DeleteUserPoolDomainInput{
		Domain:     &event.PhysicalResourceID,
		UserPoolId: &properties.UserPoolId,
	}).Send(ctx)
	if err != nil {
		awsErr, ok := err.(awserr.RequestFailure)
		if !ok || awsErr.StatusCode() != 400 || !strings.HasPrefix(awsErr.Message(), "No such domain or user pool exists") {
			return event.PhysicalResourceID, nil, errors.Wrapf(err, "could not delete the CognitoUserPoolDomain %s", event.PhysicalResourceID)
		}
	}
	return event.PhysicalResourceID, nil, nil
}

File: aws/CloudFormation/CognitoUserPoolDomain/cognito_user_pool_domain.py
import logging

import boto3
from box import Box
from crhelper import CfnResource
from schema import And, Optional, Schema

from aws.common.cfn import resource_properties, physical_resource_id
from aws.common.schema import not_empty

helper = CfnResource()
logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)

idp = boto3.client('cognito-idp')

properties_schema = Schema({
    'UserPoolId': And(str, not_empty, error='not empty string for UserPoolId'),
    'Domain': And(str, not_empty, error='not empty string for UserPoolId'),
    Optional('CustomDomainConfig', {}): {
        'CertificateArn': And(str, not_empty, error='not empty string for CertificateArn')}
})


def validate_properties(properties):
    return Box(properties_schema.validate(properties), camel_killer_box=True)


@helper.create
@helper.update
def create(event, _):
    properties = validate_properties(resource_properties(event))
    certificate_arn = properties.custom_domain_config.certificate_arn
    if certificate_arn is None:
        out = idp.create_user_pool_domain(
            Domain=properties.domain,
            UserPoolId=properties.user_pool_id
        )
    else:
        out = idp.create_user_pool_domain(
            Domain=properties.domain,
            UserPoolId=properties.user_pool_id,
            CustomDomainConfig=properties.custom_domain_config
        )
    cloudfront_domain = out.get('CloudFrontDomain')
    if cloudfront_domain is None:
        cloudfront_domain = ''
        domain = properties.domain + '.auth.' + idp._client_config.region_name + '.amazoncognito.com'
    else:
        domain = properties.domain

    helper.Data.update({
        'UserPoolId': properties.user_pool_id,
        'CloudFrontDomain': cloudfront_domain,
        'Domain': domain
    })

    return properties.domain


@helper.delete
def delete(event, _):
    properties = validate_properties(resource_properties(event))
    delete_user_pool_domain(properties)
    return physical_resource_id(event)


def delete_user_pool_domain(properties):
    try:
        idp.delete_user_pool_domain(
            Domain=properties.domain,
            UserPoolId=properties.user_pool_id
        )
    except idp.exceptions.InvalidParameterException:
        pass


def handler(event, context):
    logger.info('event: %s', event)
    helper(event, context)
